{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img style=\"float: left; margin-right:20px;\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/trifusion-icon-64.png\"><p style=\"font-weight:bold; font-size:23px; line-height:23px; margin-bottom:50px; color:#37abc8;\">Basic conversion and concatenation of multiple alignment files.</p>\n",
    "\n",
    "Here I'll demonstrate how to apply secondary operations (individually or in combination) to the concatenation or conversion of a set of 614 alignments files with a total of 48 taxa. The covered secondary operations are:\n",
    "\n",
    "- [Collapse](#Collapse)\n",
    "- [Consensus](#Consensus)\n",
    "- Filter\n",
    "    - [Taxa filter](#Taxa-filter)\n",
    "    - [Codon position filter](#Codon-position-filter)\n",
    "    - [Gap/Missing data filter](#Gap/Missing-data-filter)\n",
    "    - [Sequence variation filter](#Sequence-variation-filter)\n",
    "- [Gap coding](#Gap-coding)\n",
    "\n",
    "Each one of these operation will be executed individually. In the end of the tutorial, a [combination of collapsing, taxa filter and gap/missing data filter operations](#Combination-of-three-secondary-operations) will also be execute to demonstrate how these can be combined.\n",
    "\n",
    "# Loading the data\n",
    "\n",
    "As already covered in a separate tutorial (see REF), alignment data can be loaded into TriFusion in three different ways. Here we will use the filechooser to load an entire directory where 614 alignments files are stored. \n",
    "\n",
    "Navigate to _Menu_ > _Open/View Data_ and click the __Open file(s)__ button. This will open the main filechooser.\n",
    "\n",
    "![TriFusion's main filechooser](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_main_filechooser_dir.png)\n",
    "\n",
    "The _input data type_ is already correctly set to __Alignment/Sequence set__, so we'll leave that as it is. Then, browse the filechooser until you find the directory containing the alignment files. In this case, all alignments are stored in a directory named \"_Version2_\". Since TriFusion supports the selection of directories (in which case all files inside the specified directory will be loaded), I will only select the \"_Version2_\" directory and click __Load & go back__ button. At the end of the data loading, a popup informs how many files were loaded.\n",
    "\n",
    "All secondary operations can be applied on top of either __Conversion__ or __Concatenation__ main operations, but in some secondary operations one of the main operations is generally more adequate, as it will be stated below.\n",
    "\n",
    "# Displaying secondary operations\n",
    "\n",
    "To display all secondary operations, click the __Show additional options__ button. This will reveal a tabbed panel, where the secondary operations are sorted into categories (with the exception of the _Formatting_ tab, which is not a secondary operation _per se_).\n",
    "\n",
    "![Secondary operations tabs](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_secondary_ops.png)\n",
    "\n",
    "# Collapse\n",
    "\n",
    "Turn __ON__ the collapse switch to activate the operation.\n",
    "\n",
    "The _Collapse_ secondary operation contains three options:\n",
    "\n",
    "- _Save in new output file_: This will save the collapse alignment in another output file, separated from the main concatenated/converted output file. Checking this option will effectively produce two output files - a main output that is only concatenated/converted and another output file with the suffix \"_collapsed\" that will be concatenaded AND collapse. For now, we will not check this option.\n",
    "\n",
    "- _Ignore missing data_: If this option is checked, sequences will be collapsed based on alignment columns that do not contain missing data and the output alignment will also contain 0% of missing data. The currently loaded data set has a fair amount of missing data, and is most likely not appropriate for collapsing using this option, so we will also leave this unchecked.\n",
    "\n",
    "- _Haplotype prefix_: Sets the prefix for the haplotypes that will appear as the taxa names in the output file. An auxiliary file with the suffix \"_haplotypes\" will also be generated when performing this operation matching the new haplotype prefix to the original taxon names. Here we can change the default value to anything, like _Haplotype_\n",
    "\n",
    "![Collapse options](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_collapse2.png)\n",
    "\n",
    "# Consensus\n",
    "\n",
    "Turn __ON__ the consensus switch to activate the operation.\n",
    "\n",
    "The consensus operation is mainly used to compress multiple sequences in an alignment into one representative sequence. While it can be done on top of the __Concatenation__ main operation, what this will do is concatenate all 614 alignments into a single concatenated one and then create a consensus of that large alignment. However, in the majority of the cases, users are more interested in creating a consensus sequence for each input alignment. With this in mind, this secondary operation will be done with the __Conversion__ main operation selected for this tutorial.\n",
    "\n",
    "The _Consensus_ secondary operation contains three options:\n",
    "\n",
    "- _Save in new output file_: This will save the consensus alignment in another output file, separated from the main concatenated/converted output file. Checking this option will effectively produce two output files - a main output that is only concatenated/converted and another output file with the suffix \"_consensus\" that will be concatenaded AND consensus. For now, we will not check this option.\n",
    "\n",
    "- _Save consensus in a single file_: This option can be check to merge all consensus from each input alignment in a single file. In this case, if this option is left unchecked, 614 output files will be created using this option, each with a single representative consensus sequence of the corresponding alignment. However, here we are more interested in merging all consensus sequences in a single file that will be later provided to Blast2GO for functional annotation. So check this option.\n",
    "\n",
    "- Consensus variation handing_: Select how you would like to handle variation within each alignment. The appropriate choice is highly dependent on subsequent analyses. In our case, since we want to create a dataset for Blast2GO and our alignment data is fairly variable, we'll select the _First sequence_ value, where the first sequence of each alignment is selected as a representative. \n",
    "\n",
    "![Consensus options](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_consensus.png)\n",
    "\n",
    "# Taxa filter\n",
    "\n",
    "Turn __ON__ the filter switch to activate the operation. Then click the __Set filters__ button for the _Taxa filter_ option and turn __ON__ the switch n the popup as well.\n",
    "\n",
    "The _Taxa filter_ operation allows users to filter entire alignments if they contain or exclude a given set of taxa. Here, we will create a fictious case where we are interested in concatenating only alignments that contain at least all taxa with names beggining on a \"C\". \n",
    "\n",
    "![Initial taxa filter dialog](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_taxa_filter_new.png)\n",
    "\n",
    "The filter mode sets whether the alignments should be filtered if they _contain_ or _exclude_ the taxa group. By default, it is set to _Containing_, so we'll leave that unchanged.\n",
    "\n",
    "As you can see, there are no taxa groups yet defined so we'll need to create a new one. Click the __Set taxa group__ button to start the dataset group creation process and then click the __Set manually__ button. Here, select the desired taxa with names starting with the letter \"C\" and save the taxa group by clicking __Ok__.\n",
    "\n",
    "![Taxa group definition](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_concatenation_taxa_group.png)\n",
    "\n",
    "Once the group has been created, it will be automatically selected in the _Taxa filter_ dialog. Additional groups can be created in the same way. When multiple groups have been defined, they can be selected by clicking the __Use taxa group__ button, and then the desired group. \n",
    "\n",
    "![Taxa filter ialog with group defined](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_taxa_filter_with_group.png)\n",
    "\n",
    "When you are happy with the taxa filter settings, click the __Ok__ button. If the _Taxa filter_ switch was turned __ON__, the button of the _Taxa filter_ option should change to __Filters ON__.\n",
    "\n",
    "![Active taxa filter](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_taxa_filter_active.png)\n",
    "\n",
    "\n",
    "# Codon position filter\n",
    "\n",
    "Turn __ON__ the filter switch to activate the operation. Then click the __Set filters__ button for the _Codon position filter_ option and turn __ON__ the switch on the popup as well.\n",
    "\n",
    "The _Codon position filter_ operation allows you to remove certain codon positions from the output alignment. Consequently, this option is only available for nucleotide sequences. In many nucleotide alignments it is common to remove the third codon position, as it is generally much more variable and could introduce a substantial amount of phylogenenetic noise. However, this option removes the same codon positions in all input alignments. For example, if you load 10 alignments in TriFusion and exclude the 3rd codon position, you must make sure that all 10 alignments start in the 1st codon position. However, if all alignments start in the 2nd codon position, for instance, removing the 3rd codon position is still possible in TriFusion, by excluding the 2nd positions (which will actually correspond to the 3rd positions in the alignment). \n",
    "\n",
    "To exclude a given codon position, simply toggle the corresponding button off. Included position button always have a blue background.\n",
    "\n",
    "![Codon position filter dialog](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_codon_filter.png)\n",
    "\n",
    "# Gap/Missing data filter\n",
    "\n",
    "Turn __ON__ the filter switch to activate the operation. Then click the __Set filters__ button for the _Gap/Missing data filter_ option and turn __ON__ the switch on the popup as well.\n",
    "\n",
    "The _Gap/Missing data filter_ allows user to filter alignment columns (_within alignment_) and/or alignments (_multiple alignments_) based on their missing data content. Both filters can be used in combination, if both _within alignment_ and _multiple alignments_ checkboxes are active, or only one of them.\n",
    "\n",
    "![Missing data filter initial dialog](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_missing_filter_initial.png)\n",
    "\n",
    "In this example, we will filter both aligment columns and alignment files, so both checkboxes will remain active. Within an alignment, columns can be filtered depending on the amount of gaps or missing data. Gaps refer to the usual gap symbol (\"-\" or \"?\") while missing data refers to the sum of gap symbols __AND__ true missing data (\"N\" for nucleotides or \"X\" for proteins). These filters provide maximum threshold values in percentages, above which alignment columns are filtered. For example, if the _gap percentage allowed_ option is set to 25% and the _missing data percentage allowed_ option is set to 50%, then alignment columns with more than 25% of gaps OR more than 50% of gaps + true missing data are filtered. \n",
    "\n",
    "In our case, we are interested in producing an output matrix that contains no missing data, so we will set both sliders to 0%.\n",
    "\n",
    "Concerning the _multiple alignments_ option, we will be more relaxed. We'll set the slider to 25%, which means \n",
    "that only alignments with more than 25% of the total dataset taxa (12 out of 48 in this case) will be further process.\n",
    "\n",
    "![Missing filter dialog with final settings](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_missing_filter_final.png)\n",
    "\n",
    "When you are happy with the gap/missing data filter settings, click the __Ok__ button. If the _Gap/Missing data filter_ switch was turned __ON__, the button of the _Gap/Missing data filter_ option should change to __Filters ON__.\n",
    "\n",
    "![Missing filter active](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_missing_filter_active.png)\n",
    "\n",
    "\n",
    "# Sequence variation filter\n",
    "\n",
    "Turn __ON__ the filter switch to activate the operation. Then click the __Set filters__ button for the _Sequence variation filter_ option and turn __ON__ the switch on the popup as well.\n",
    "\n",
    "The _sequence variation filter_ allows users to filter alignment files based on the amount of sequence variation. The two supported types of sequence variation are _variable sites_ and _informative sites_. The different between these types is that _variable sites_ includes all columns with at least one variant, while _informative sites_ only includes variable columns where at least one alternative allele has two or more copies.\n",
    "\n",
    "![Variation filter initial dialog](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_variation_filter_initial.png)\n",
    "\n",
    "Here, you can specify multiple combination of maximum and minimum values for each variation type. When a checkbox os left inactive, it is assumed that there is no boundary for that specific value. For instance, lets filter our alignments so that only alignments with at least 2 informative sites are processed. To achieve this, check the _Minimum_ box of the _informative sites_ option and set it to 2, but leave the _Maximum_ box unchecked. \n",
    "\n",
    "![Variation filter var1](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_variation_filter_var2.png)\n",
    "\n",
    "If you would like to set an upper limit to the number of informative sites, just check the _Maximum_ box and set a number higher than 2. In this case, let's put an upper limit of 10 informative sites.\n",
    "\n",
    "![Variation filter var2](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_variation_filter_var3.png)\n",
    "\n",
    "It is also possible to mix both types of sequence variation. For instance, we may want to filter alignments with more than 2 informative sites and less than 5 variable sites. \n",
    "\n",
    "![Variation filter var3](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_variation_filter_var4.png)\n",
    "\n",
    "However, note that certain combination are redundant. For instance, if you set a mininimum of informative sites to 2, setting a minimum variable sites to 1 will have no effect on the final output. \n",
    "\n",
    "When you are happy with the sequence variation filter settings, click the __Ok__ button. If the _Sequence variation filter_ switch was turned __ON__, the button of the _Sequence variation filter_ option should change to __Filters ON__.\n",
    "\n",
    "![Variation filter active](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_variation_filter_active.png)\n",
    "\n",
    "# Gap coding\n",
    "\n",
    "Turn __ON__ the gap coding switch to activate the operation. \n",
    "\n",
    "The _Gap coding_ enables the codification of gaps as a binary matrix that is appended to the final of the alignment matrix. __This option is available only when the Nexus format is the only output format selected__. Currently, it contains a single available option:\n",
    "\n",
    "- _Save in new output file_: This will save the alignment with coded gaps in another output file, separated from the main concatenated/converted output file. Checking this option will effectively produce two output files - a main output that is only concatenated/converted and another output file with the suffix \"_gcoded\" that will be concatenaded AND contains the gaps coded. For now, we will not check this option.\n",
    "\n",
    "The _Gap coding method_ is currently restricted to the one described in [Simmons and Ochotenera 2000](https://www.ncbi.nlm.nih.gov/pubmed/12118412), however additional methods are expected to be added in future releases.\n",
    "\n",
    "\n",
    "# Combination of three secondary operations\n",
    "\n",
    "Until now, we only dealt with the activation and usage of individual secondary operations. However, many of these  operations can fit rather naturally in combination. Here I'll demonstrated how a dataset of 614 alignments with 48 taxa can be concatenated, collapsed and filtered in a single run, with the condition that the collapsed alignment has to be generated in an independent alignment file.\n",
    "\n",
    "After loading the data, select the __Concatenation__ main operation in the __Process__ screen. To keep things simple, let's leave the _Data set_ options in the default values, select only the _Nexus_ output format and provide an output file name (here it will be <i>my_concatenation</i>).\n",
    "\n",
    "![General options for Concatenation operation](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_general_opts_secops.png)\n",
    "\n",
    "### Setting up collapse operation\n",
    "\n",
    "Open the secondary operations tabbed menu by clicking the __Show additional options__ button, click on the _Collapse_ tab and turn the switch __ON__. \n",
    "\n",
    "Since we want to save the collapsed alignment in a separated output that is independent of the remaining operations, we'll check the _Save in new output file_ box. Our data set contains a fair amount of missing data, so we'll leave the _Ignore missing data_ box uncheked. Finally, we can leave the haplotype prefix in its default _Hap_ value. \n",
    "\n",
    "![Collapse settings](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_seccombination_collapse.png)\n",
    "\n",
    "### Setting up taxa filter\n",
    "\n",
    "Here we are interested creating an output dataset that __excludes__ alignments that contains any taxon whose name starts with the letter \"C\".\n",
    "\n",
    "Click on the _Filter_ tab and turn the switch __ON__. Then, click on the __Set filters__ button for the _Taxa filter_ option and active the switch in the popup. Change the _Filter mode_ to _Exclude_ and then click on the __Set taxa group__ button to define the new taxa group. \n",
    "\n",
    "![Taxa filter initial dialog](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_seccomb_taxa_filter_initial.png)\n",
    "\n",
    "Let's manually create a taxa group with all taxa names that start with the letter \"C\" by clicking the __Set manually__ button. \n",
    "\n",
    "![Taxa group definition](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_concatenation_taxa_group.png)\n",
    "\n",
    "Once the group has been created, check that the this group is correctly selected in the _Taxa filter_ dialog.\n",
    "\n",
    "![Taxa filter final](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_seccomb_taxa_filter_final.png)\n",
    "\n",
    "If all checks out, click __Ok__ and the button of the _Taxa filter_ option should now display __Filters ON__\n",
    "\n",
    "### Setting up missing data filter\n",
    "\n",
    "Here we are interested in filtering ONLY alignments that contain less than 50% of the total taxa in the dataset. Since we are not interested in the _within alignment_ filtering, let's uncheck this box and set the _Multiple alignments_ slider to 50%. \n",
    "\n",
    "![Missing filter final](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_seccomb_missing_filter_initial.png)\n",
    "\n",
    "Then select the __Ok__ button, and both the _Taxa filter_ and _Gap/Missing data filter_ buttons should now display __Filters ON__\n",
    "\n",
    "![Filters final](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_seccomb_filter_final.png)\n",
    "\n",
    "### Checking selected options\n",
    "\n",
    "All currently active options can be viewer by clicking the __View Queue__ button at the bottom of the __Process__ screen. This will open the _Menu_ sidepanel and show that:\n",
    "\n",
    "- The main operation if __Concatenation__;\n",
    "- There are two active secondary operations: __Collapse__ and __Filter__;\n",
    "- The Nexus output format is the only selected;\n",
    "- There are two expected output files: The main output, <i>my_concatenation</i>, and the separate output file that will only contain the result of the concatenation and collapse operations, <i>my_concatenation_collapse</i>. \n",
    "\n",
    "![Operations queue](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_seccomb_queue.png)\n",
    "\n",
    "### Execution\n",
    "\n",
    "If everything checks out, click the __Execution__ button at the bottom of the __Process__ screen to show the  small popup that displays a summary of the process execution and then click the __Execute__ button to begin the execution.\n",
    "\n",
    "![Execution summary](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_seccomb_execution.png)\n",
    "\n",
    "At the end of the execution, a filter report will appear showing the number of alignments that were filtered by the active filters. Since we only actived two of the four filters that can remove alignments from the final output, the values for the other two filters display a _Not applied_ message. For the active filters, the number of alignments removed due to that filter is displayed. In this case, no alignment was removed from the _Gap/Missing data_ filter (it seems all alignments already contained more than 50% of the total taxa) and 84 alignments were removed by the _Taxa filter_.\n",
    "\n",
    "![Filter report](https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/process_seccomb_filter_report.png)\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
