{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img style=\"float: left; margin-right:20px;\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/trifusion-icon-64.png\"><p style=\"font-weight:bold; font-size:30px; color:#37abc8; margin-bottom:50px;\">Export ortholog groups into protein and nucleotide sequence files</p>\n",
    "\n",
    "Here I'll demonstrate how to export ortholog groups from a previous __Orthology search__ operation into protein and nucleotide sequences:\n",
    "\n",
    "- [Export into protein sequences](#Export-into-protein-sequences)\n",
    "- [Export into nucleotide sequences](#Export-into-nucleotide-sequences)\n",
    "\n",
    "__Note:__ When performing an __Orthology search__ in TriFusion, protein sequence files are automatically generated for the orthology filters that were set before the run. However, in the __Explore__ section you can still modify the orthology filters and re-export the protein sequences.\n",
    "\n",
    "The first section will be a demonstration of how to load group files into TriFusion and changing the ortholog filters. If you have already done this, you can skip the next section.\n",
    "\n",
    "The data required to complete this tutorial are:\n",
    "\n",
    "- [Ortholog group files](https://github.com/ODiogoSilva/TriFusion-tutorials/raw/master/tutorials/Datasets/Ortholog_groups/groups.zip)\n",
    "- [Protein database](https://github.com/ODiogoSilva/TriFusion-tutorials/raw/master/tutorials/Datasets/10genomes/Protein_database/protein_db.zip)\n",
    "- [CDS files](https://github.com/ODiogoSilva/TriFusion-tutorials/raw/master/tutorials/Datasets/10genomes/cds/10cds.zip)\n",
    "\n",
    "# Importing group files\n",
    "\n",
    "Let's import the results from the previous search of orthologs across 10 genomes (see tutorial [Basic search of orthologs among 10 proteome files](http://nbviewer.jupyter.org/github/ODiogoSilva/TriFusion-tutorials/blob/master/tutorials/Basic%20search%20of%20orthologs%20among%2010%20proteome%20files.ipynb)). Navigate to the __Orthology__ screen, __Explore__ operation, and click the \"__+__\" button at the top left of the screen. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/explore_add_groups.png\">\n",
    "\n",
    "Go to the directory containing the group files from the corresponding __Orthology search__ operation and select the desired file.\n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/explore_group_selection.png\">\n",
    "\n",
    "Here I'll select the <i>group_3.txt</i> file. Once loaded, the basic information of the group file will be displayed for the default orthology filters (only single copy genes present in all species).\n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/explore_group_loaded.png\">\n",
    "\n",
    "However, let's change the filters for something more permissive in terms of minimum taxa representation. Click the __Change filters__ button, and change the _minimum number of taxa_ value to 5 (50% of taxa representation). Click __Ok__ and the information on the screen should be updated to something like this.\n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/explore_after_filter_change.png\">\n",
    "\n",
    "\n",
    "# Export into protein sequences\n",
    "\n",
    "First click the __Export as...__ button in the __Explore__ operation screen. This will open the export group dialog. To export the ortholog groups into protein sequence files (in Fasta format), a protein database of all input genomes must be provided. This file is automatically generated during the __Orthology search__ operation and is stored in the <i>backstage_files</i> directory, with the default name of <i>goodProteins_db</i> (This name can be change by the <i>Database name</i> option). If you have just finished an __Orthology search__ operation in the current session of TriFusion, this database file is already set. However, if you are executing a different session of TriFusion, you'll need to provide this file.\n",
    "\n",
    "Click the __Protein sequences__ button. This will make the _Protein database_ base option available. To search and select the database file, click the __Select...__ button. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/explore_export_protein.png\">\n",
    "\n",
    "Note that I navigated to the results directory of my previous ortholog search and then to the <i>backstage_files</i> directory. Since I did not change the _Database name_ value in TriFusion, I have a <i>goodProteins_db</i> file in this directory. Select this file and click __Save__. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/explore_select_protein_db.png\">\n",
    "\n",
    "You'll notice that the _Protein database_ button changed in accordance to the name of the protein database file. Finally, to export the ortholog groups click the __Export__ button. Select or create a directory where the new files will be generated and then click __Ok__. At the end of the export operation, a success popup should appear informing the number of ortholog groups exported. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/explore_protein_export_success.png\">\n",
    "\n",
    "\n",
    "# Export into nucleotide sequences\n",
    "\n",
    "__NOTE:__ To export ortholog groups, a working executable of USEARCH is required. See the [Installation of third party software (USEARCH and MCL)](http://nbviewer.jupyter.org/github/ODiogoSilva/TriFusion-tutorials/blob/master/tutorials/Installation%20of%20third%20party%20software%20%28USEARCH%20and%20MCL%29.ipynb) tutorial.\n",
    "\n",
    "First click the __Export as...__ button in the __Explore__ operation screen. This will open the export group dialog. To export the ortholog groups into nucleotide sequence files (in Fasta format), a protein database AND cds/transcript files must be provided.\n",
    "\n",
    "Click the __Nucleotide sequences__ button. This will make available the _Protein database_ and _CDS database_ options. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/explore_export_nuc.png\">\n",
    "\n",
    "Refer to the [previous section](#Export-into-protein-sequences) on how to set the protein database file. After setting this file, the cds/transcripts that correspond to the proteomes used during the __Orthology search__ operation, must be also provided. You can have an individual cds/transcript file for each species, or concatenate all files into a single master file. Click the __Select...__ button of the _CDS database_ option and search for the cds/transcript files.\n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/explore_select_cds.png\">\n",
    "\n",
    "Here, I have the CDS and transcript data for each of the 10 species in their respective individual files. Select them all with shift + click and click __Save__. You should notice that the _CDS database_ button changed in accordance to the number of files select, which is 10 in this case.\n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/explore_export_nuc_ready.png\">\n",
    "\n",
    "With both the protein database and cds/transcript files selected, we are ready to begin the ortholog export. Click the __Export__ button and select the directory where you want to generate the nucleotide sequence files. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/export_nuc_output_dir.png\">\n",
    "\n",
    "At the end of the export operation, a success popup should appear informating the number of \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/export_nucleotide_success.png\">\n",
    "\n",
    "__Note on the sequences that could not be retrieved:__ TriFusion converts groups into nucleotide sequences by searching the proteins from the main output of the __Search__ operation in CDS/transcript databases provided by the user. The reason why this search is done instead of simply looking for sequence headers that are the same in the protein and nucleotide databases is that most of the time there is no such cross reference. Therefore, TriFusion creates two different databases and then uses _USEARCH_ to search for perfect hits between the protein and nucleotide sequences. This ensures that the nucleotide sequences correspond exactly to the proteins referenced during the __Orthology search__ operation. However, even with this method, some nucleotide sequences may be absent (for several reasons) from the databases. Fortunately, this represents only a small minority of the cases. In this example, 641 protein sequence had no match in the nucleotide databases provided by the user, which represents only 2.8% of the total dataset. In most cases, this occurs only on a limited number of species but in any case, make sure that the proteome and CDS/transcript files correspond to the same version of the genome sequencing project. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
