{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img style=\"float: left; margin-right:20px;\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/trifusion-icon-64.png\"><p style=\"font-weight:bold; font-size:22px; color:#37abc8; margin-bottom:50px;\">Import and explore multiple orthology search results</p>\n",
    "\n",
    "Here, I'll show how to import the output of multiple orthology search operations and how to explore the data vi the _Explore_ section of TriFusion's __Orthology__ module. The input data that is used for this tutorial can be downloaded [here](https://github.com/ODiogoSilva/TriFusion-tutorials/raw/master/tutorials/Datasets/Ortholog_groups/groups.zip).\n",
    "\n",
    "We will cover how to:\n",
    "\n",
    "- [Load data](#Load-data)\n",
    "- [Change the active group file](#Change-the-active-group-file)\n",
    "- [Change the orthology filters](#Change-the-orthology-filters)\n",
    "- [Compare ortholog groups among group files](#Compare-ortholog-groups-among-group-files)\n",
    "- [Further investigate individual group files](#Further-investigate-individual-group-files)\n",
    "- [Generate full exploration report for individual group files](#Generate-full-exploration-report-for-individual-group-files)\n",
    "\n",
    "# Load data\n",
    "\n",
    "The input data of the _Explore_ operation of the __Orthology__ module are the group files that are generated at the end of the orthology search operation. These are simple text files that contain the definition of an ortholog group in each line. A typical group file should start with something like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "My_group1000: Aurde|1093705\tAurde|1159865\tAurde|1196783\tAurde|76841\tPhbla|138389 (...)\n",
    "My_group1001: Phbla|138268\tPhbla|139062\tPhbla|145188\tPhbla|145348\tPhbla|145707 (...)\n",
    "My_group1002: Phbla|160931\tPhbla|137818\tPhbla|137820\tPhbla|137828\tPhbla|142143 (...)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you are loading group files from previous orthology search runs, they will be found inside the specified output directory, in the <i>Orthology_results</i> directory. \n",
    "\n",
    "To load the data, navigate to the __Orthology__ screen and click the _Explore_ tab button on the left of the screen.\n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_screen.png\">\n",
    "\n",
    "Then click the \"__+__\" button on the top left of the screen to open the filechooser. Navigate to the directory containing the group files and then select the desired files. In this case, we will selected the three group files generated in a previous search operation that were performed with inflation values 2, 3 and 4. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_group_files.png\">\n",
    "\n",
    "Once the group files are loaded into TriFusion, several descriptive statistics will populate the screen.\n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_cards.png\">\n",
    "\n",
    "To the left, the loaded group files are listed under the _Group file(s)_ section, where they can be selected to visualize the statistics specific to that group. They can also be removed by clicking the trash bin red button. On the remaining of the screen, general statistics and information on the filtered orthologs are presented for the currently selected group file. \n",
    "\n",
    "The _General information_ section informs the total number of proteins, taxa and ortholog groups contained in the group file. \n",
    "\n",
    "Below, in the _Filtered orthologs_ section, the number of orthologs after applying the filters is displayed in gaussian plots. The values displayed are for the default ortholog filter, wich are set to single copy genes (maximum gene copies of 1) and with all taxa present (minimum number of taxa equal to the number of taxa). \n",
    "\n",
    "In our case, we can see that the <i>group_2.txt</i> file contains around 1.5M proteins for 10 taxa, clustered in 20 268 ortholog groups. From these 20k ortholog groups, 1 934 passed the species filter (minimum number of taxa), 9 141 passed the gene filter (maximum number of gene copies) and 1 132 passed both filters. This indicates that the species filter is the major limiting factor in the final number of ortholog groups.\n",
    "\n",
    "# Change the active group file\n",
    "\n",
    "To change the active group file, simply click the group button in the _Group file(s)_ list section in the top left of the screen. Let's change the active group file to the <i>group_3.txt</i> file.\n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_change_group.png\">\n",
    "\n",
    "As you can see, the numbers of total and filtered ortholog groups changed slightly, which is a result from using different inflation values during the search operation. \n",
    "\n",
    "# Change the orthology filters\n",
    "\n",
    "A common procedure during the exploration of the orthology search results is the modification of the ortholog filters. To change the filters, click the __Change filters__ button in the bottom of the _Filtered orthologs_ section. This will open the ortholog filters dialog where you can change the maximum number of gene copies and minimum number of taxa for the ortholog groups. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_expore_change_filters.png\">\n",
    "\n",
    "Let's maintain the gene copy filter and still only allow single copy genes, but relaxed the minimum number of taxa to half of the data set (5). The _Apply filter to all group files_ check box will also remain active to update all group files with the new filter. When all filters are set, click the __Ok__ button to update.\n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_after_filter.png\">\n",
    "\n",
    "After the application of the new filters, you can see that the number of filtered orthologs changed. The number of final orthologs for the <i>group_3.txt</i> file almost trippled when we relaxed the number of minimum taxa per ortholog group. You can also see that the filter values were updated at the bottom of the _Filtered orthologs_ section. \n",
    "\n",
    "# Compare ortholog groups among group files\n",
    "\n",
    "To easily compare the number of total and filtered ortholog groups among different group files, you can check he boxes to the left of the group files in the _Group file(s)_ list section. To select/deselect all group files, you can also check the top checkbox. Here, let's compare all group files by selecting all and then clicking on the __Compare__ button. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_compare.png\">\n",
    "\n",
    "This will bring you to a plotting screen, where a bar plot will be displayed with the number of total and filtered ortholog groups for each group file. You can interact with the plot by pressing the left mouse buttn and dragging the plot. You can also zoom in and out using _Ctrl + mouse wheel_ or by clicking the corresponding buttons on the right side panel. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_compare_plot.png\">\n",
    "\n",
    "At the top of the screen, you can see the currently active filters, which are the same we set in the previous section. You can change the filter values using the sliders. Let's try to relaxed even further the minimum number of taxa to 2. After changing the slider value, you can see that the refresh button turned red, which means that you have set different filter. To update the plot, click the refresh button. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_compare_changed_filter.png\">\n",
    "\n",
    "After clicking the refresh button, the plot values will be updated. You can see now that the total number of orthologs is almost 10k for all group files and that there is almost no different between the gene filtered and final ortholog groups. Indeed, we can see that the final number of orthologs does not deviate much between group files (range between 9 141 and 10 620). \n",
    "\n",
    "You can also change which type of ortholog groups are displayed by ticking the check boxes in the _Display_ section on the top right of the screen. Let's visualize only the total and final number of orthologs. To accomplish this, uncheck the _Gene filter_ and _Species filter_ boxes. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_compare_change_display.png\">\n",
    "\n",
    "At any time, you can export the current plot in figure or table format by clicking the __Export as graphics__ or __Export as table__ buttons, respectively, in the right side panel. You can also export the ortholog groups into another group file, protein or nucleotide sequences by clicking the __Export group__ button (See tutorial [Export ortholog groups into protein and nucleotide sequence files](http://nbviewer.jupyter.org/github/ODiogoSilva/TriFusion-tutorials/blob/master/tutorials/Export%20ortholog%20groups%20into%20protein%20and%20nucleotide%20sequence%20files.ipynb)). \n",
    "\n",
    "# Further investigate individual group files\n",
    "\n",
    "Individual group files can also be further visually explored using the ploting tools under the _Graphical visualization_ section in the bottom left of the screen. Graphical visualization options are sorted into _Species focused exploration_ and _Ortholog focused exploration_. Clicking on either option will present a drop down menu where specific ploting options are available. When one of these options is selected, a short description is shown below. Let's investigate the taxa coverage of the currently active group file, by selecting the _Species focused exploration_ and the _Taxa coverage_ plot option. Then, click on the __Generate plot__ button. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_graphical_visualization.png\">\n",
    "\n",
    "This will open a plot screen akin to the one displayed when comparing different group files. In this specific plot you can see, for each taxa, the proportion of ortholog groups where they are present (dark blue) or missing (light blue). The horizontal dashed red line represents the data set mean. In the top right of the screen, under the _Summary_ section, you can see the total (red) and filtered (green) number of ortholog groups and taxa that are being used to generated the plot. In this case, a total of 21 777 ortholog groups across 10 taxa are being used.\n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_taxa_coverage.png\">\n",
    "\n",
    "The plot can be interacted with by clicking and dragging and by zooming in and out. In the header of the screen, the ortholog filters can be changed. Let's change the filter setting so that only single copy genes with at least 5 taxa represented are considered.  When the filters are modified, the refresh button should turn red and must be clicked in order to update the plot.\n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_taxa_coverage_change_filter.png\">\n",
    "\n",
    "After the plot is updated, you can see that the values in the _Summary_ section of the header have also updated. This plot is now being generated with 4 688 ortholog groups across 10 taxa. We can also see that,  using these filter values, all taxa have a pretty decent proportion of available data. However, you have also the option to remove specific taxa from this analyses, by clicking the filter taxa button in the header above the refresh button. Clicking this button will display all taxa listed. These can be toggled in or out by clicking the respective buttons. For exampled, let's remove the last two taxa, _Thite_ and _crneo_, by clicking them once. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_taxa_coverage_remove_taxa.png\">\n",
    "\n",
    "As you can see, the bars of the removed taxa are no longer in the plot and the numbers in the _Summary_ section of the header were updated to 8 active taxa. \n",
    "\n",
    "As in the compare groups plot screen, all plots in the _Graphical visualization_ section can be exported into figures or table formats by clicking the __Export as graphics__ or __Export as table__ buttons, respectively. \n",
    "\n",
    "# Generate full exploration report for individual group files\n",
    "\n",
    "All plotting options in the _Graphic visualization_ section can be automatically generated into a HTML file by clicking the __Generate full report__ button at the bottom of the _Explore_ screen. \n",
    "\n",
    "<img width=\"90%\" src=\"https://raw.githubusercontent.com/ODiogoSilva/TriFusion-tutorials/master/tutorials/images/orto_explore_full_report.png\">\n",
    "\n",
    "Then select the directory where the report will be generated. In that directory, an HTML file will be created where all plots will be visualized for the currently set ortholog filters."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
